### PR TITLE
ENH: stats.chatterjeexi: add xi correlation function

### LINKED ISSUE #21453: ENH: Add support for Xi Correlation in scipy
### Is your feature request related to a problem? Please describe.

When trying to identify non-linear and non-monotonic correlations, Xi correlation is a good alternative to person, spearman and Kendall correlation. More information about the Xi correlation here:

1. https://towardsdatascience.com/a-new-coefficient-of-correlation-64ae4f260310
2. https://arxiv.org/pdf/1909.10140
3. https://readmedium.com/en/https:/towardsdatascience.com/a-new-coefficient-of-correlation-64ae4f260310

@article{chatterjee2021new,
  title={A new coefficient of correlation},
  author={Chatterjee, Sourav},
  journal={Journal of the American Statistical Association},
  volume={116},
  number={536},
  pages={2009--2022},
  year={2021},
  publisher={Taylor \& Francis}
}

### Describe the solution you'd like.

Like Pearsonr() implementation, implemente Xi correlation implementation.

### Describe alternatives you've considered.

Other alternatives are existing Python libraries doing the implementation, but considering the simplicity of the correlation measure, I do not see any reason to group this within the scipy library

### Additional context (e.g. screenshots, GIFs)

https://arxiv.org/abs/1909.10140

### PR DESCRIPTION
**Pull request by mdhaber**
#### Reference issue
Closes gh-21453

#### What does this implement/fix?
This adds a function that computes the $\xi$ correlation coefficient and performs an (asymptotic or permutation) test of independence.

#### Additional information
- `alternative` could be added, but two-sided and `alternative='less'` test seem unlikely to be useful
- It's arguable that ties should be automatically detected by using `_rankdata`, and this should inform the choice of statistic calculation method.  However, the p-value calculation is supposed to be based on whether the distribution underlying `y` is continuous or not - not whether there actually *are* ties. So it seems like we should be able to use `y_continuous` for both. Perhaps automatic detection (and maybe a `metod='auto'`) can be added in a follow -up.
- When there are ties in `x`, they are broken at random (an option suggested by the paper). Perhaps we can do something smarter in a follow-up, but let's leave it alone for now.
- `too_small=1` doesn't seem to be working for multidimensional arrays
- Should we document this as the $\xi$ statistic (using the Greek letter)?

### CODE REVIEW

**Comment by mdhaber** [comment]
@matbmeijer here is a PR for the xi correlation statistic and corresponding hypothesis test. If you're still interested, please take a look and send an email to [the forum](https://discuss.scientific-python.org/c/contributor/scipy/32) about it. If there is no longer interest, let us know if we should close the PR and issue.

File: scipy/stats/_correlation.py
Comment (fancidev) [inline scipy/stats/_correlation.py:118]: My understanding from the paper is that if Y is continuous, the (asymptotic) standard error has a closed form, and if Y is not continuous, the standard error depends on the distribution of Y which is in turn estimated (consistently) from the input Y data.

For large sample size, consistency should be all that is needed. For moderate sample size, the closed form formula might be more efficient (in statistical sense). For small sample, the asymptotic distribution might not be so accurate anyway, so a little more noise in the standard error probably doesn’t matter.

File: scipy/stats/_correlation.py
Comment (fancidev) [inline scipy/stats/_correlation.py:131]: By “as high as” do you mean the magnitude? (i.e. “two-way”?)

File: scipy/stats/_correlation.py
Comment (fancidev) [inline scipy/stats/_correlation.py:132]: It might be helpful to add a note somewhere to mention that the correlation is not “symmetric”, that is xi(x, y) != xi(y, x) in general.

File: scipy/stats/_correlation.py
Comment (fancidev) [inline scipy/stats/_correlation.py:171]: I find the phrase “assumed with little effect on the results” slightly confusing, as the continuity of Y is not *assumed* but *known* a priori, and whether it impacts the p-value much probably depends mainly on the sample size (being large).

File: scipy/stats/_correlation.py
Comment (fancidev) [inline scipy/stats/_correlation.py:173]: Following my comment in the code, I think Y_continuous should affect the p-value and not the statistic value. (Here you obtained identical statistic values because the input data has no ties, so even with the current logic of the code the “if” branch and the “else” branch produce the same result.)

File: scipy/stats/_correlation.py
Comment (fancidev) [inline scipy/stats/_correlation.py:184]: Theorem 1.1 says the statistic converges (in large sample) to a constant between 0 and 1. This does *not* require the null hypothesis but only requires the samples to contain independent observations. Remark (9) following Theorem 1.1 further elaborates this and mentions that negative xi_correlation, bar noise, can only indicate non-independent observation.

Therefore I’m inclined toward a “two-way” p-value, though I’m not completely sure. But I guess “alternative=less” is out of the way.

**Review by fancidev — COMMENTED** [review]
I think this is a very interesting addition and I can’t wait to try it out once it’s added!

> When there are ties in x, they are broken at random (an option suggested by the paper). Perhaps we can do something smarter in a follow-up, but let's leave it alone for now.

To be accurate, the current code picks an unspecified order (subject to quick sort), but not a “random” order. I think this is good for reproducibility. The paper suggests to alternatively average over the possible arrangements; I guess it could be addressed in follow up if deemed necessary. (Would it be good to mention that in the doc btw?)

> Should we document this as the ξ
statistic (using the Greek letter)?

It seems the [kendalltau](https://docs.scipy.org/doc/scipy/reference/generated/scipy.stats.kendalltau.html) function is documented using the word tau. So I guess documenting using xi would be acceptable.

A few more details below for discussion.

File: scipy/stats/_correlation.py
Comment (mdhaber) [reply scipy/stats/_correlation.py:63]: I'll take a look at the version of the paper I used and see if it's different. Otherwise, this was an accidental omission.

File: scipy/stats/_correlation.py
Comment (mdhaber) [reply scipy/stats/_correlation.py:118]: Was there anything to change here, or just confirming understanding?

File: scipy/stats/_correlation.py
Comment (mdhaber) [reply scipy/stats/_correlation.py:131]: Yes to the update - it's a one-way "greater than" test as written. I put a comment about that somewhere. There could conceivably be an `alternative` parameter, but the statistic is a bit unusual - you can get negative values, but not with "anti-correlation" (which is just correlation as far as the statistic is concerned) or stochastic Independence. AFAICT, the data has to perversely avoid the appearance of correlation to give a highly negative value.

File: scipy/stats/_correlation.py
Comment (mdhaber) [reply scipy/stats/_correlation.py:171]: I can adjust it. I shouldn't have written that.

File: scipy/stats/_correlation.py
Comment (mdhaber) [reply scipy/stats/_correlation.py:173]: Yup. I know. Let's close this comment though and discuss it above.

File: scipy/stats/_correlation.py
Comment (mdhaber) [reply scipy/stats/_correlation.py:184]: Yes. That's why I went this way. Thought it was worth mentioning as something to consider. Let's make sure the reviewing maintainer is on the same page.

**Comment by mdhaber** [comment]
> I think this is a very interesting addition and I can’t wait to try it out once it’s added!

In that case, would you like to make a post on the mailing list and help find a maintainer to review? It looks like the OP does not intend to do that.

**Comment by mdhaber** [comment]
Thanks for posting @fancidev. I suppose `chatterjeexi` would be most consistent with existing names `kendalltau`, `pearsonr`, and `spearmanr` for the correlation coefficients.

**Comment by fancidev** [comment]
> In that case, would you like to make a post on the mailing list and help find a maintainer to review? It looks like the OP does not intend to do that.

Done. https://discuss.scientific-python.org/t/new-function-scipy-stats-xi-correlation/1498/1

> I suppose `chatterjeexi` would be most consistent with existing names `kendalltau`, `pearsonr`, and `spearmanr` for the correlation coefficients/tests.

Sounds good to me!

> One additional extension is adding confidence intervals. This can be added as a method of the result object as usual.

Good point!

**Review by tupui — APPROVED** [review]
Looks good and follows the paper.

@fancidev did you have more comments as you had a close look at this? Otherwise please approve if you are ok with this.

**Comment by mdhaber** [comment]
Thanks @tupui. If you'd reply to or resolve the open comments, I'd appreciate it. Some of those comments were left open for the reviewing maintainer to decide. 

Also, it looks like we had converged on renaming to `chatterjeexi` to be consistent with the names of all the existing correlation functions: `kendalltau`, `pearsonr`, and `spearmanr`. Should I push that change?

**Comment by tupui** [comment]
It's a bit ugly but yeah I suppose we should do this 👍😅

File: scipy/stats/_correlation.py
Comment (tupui) [reply scipy/stats/_correlation.py:194]: Sounds sensible and there is a precedent too 👍

**Comment by tupui** [comment]
Getting this in. It went on the forum and we left that long enough up there. Thanks again Matt!

**Pull request merged**
